### PR TITLE
feat: Allow duplicates to be dealt with in parse_survey

### PR DESCRIPTION
**Pull request by jamieRowen**
Looks to address #74 by allowing `parse_survey` to proceed even if duplicate data are found.
The existing solution errors as a result of `{assertthat}` and has no escape hatch.

This PR adds a parameter to `parse_survey`, `fix_duplicates`, which defaults to the same error behaviour as is currently present with an expanded message. But also allows the user to specify that duplicates are kept or dropped as part of the processing of survey responses. In each case if duplicates are found a warning will be presented.

- test: Add tests for functions dealing with duplicates
- default is to retain the error behaviour before

### CODE REVIEW

File: R/assemble.R [hunk 1]
```diff
@@ -34,14 +40,19 @@ parse_survey <- function(surv_obj, oauth_token = get_token(), ...) {
...
+  # to deal with this add parameter fix_duplicate where default behaviour is to error, but
+  # can be set to allow the function to continue and return.
+  fix_duplicates = match.arg(fix_duplicates)
+  if (fix_duplicates == "error") {
+    x = duplicate_error(x)
+  } else if (fix_duplicates == "keep") {
+    x = duplicate_keep(x)
+  } else {
+    x = duplicate_drop(x)
+  }
```
Comment (mattroumaya) [inline R/assemble.R:54]:
```suggestion
  fix_duplicates <- match.arg(fix_duplicates)
  if (fix_duplicates == "error") {
    x <- duplicate_error(x)
  } else if (fix_duplicates == "keep") {
    x <- duplicate_keep(x)
  } else {
    x <- duplicate_drop(x)
  }
```

File: R/assemble.R [hunk 2]
```diff
@@ -241,3 +252,55 @@ de_duplicate_names <- function(x) {
   )
   x
 }
+
+
+# does a data frame contain duplicate rows
+# @param x a data.frame
+# @return logical, TRUE if there are any duplicates in the data.frame
+contains_duplicates = function(x) {
```
Comment (mattroumaya) [inline R/assemble.R:260]:
```suggestion
contains_duplicates <- function(x) {
```

File: R/assemble.R [hunk 3]
```diff
@@ -241,3 +252,55 @@ de_duplicate_names <- function(x) {
...
+  sum(find_duplicates(x)) > 0
+}
+
+# @param x a data.frame
+# @return a logical vector of the duplicated rows
+find_duplicates = function(x) {
```
Comment (mattroumaya) [inline R/assemble.R:266]:
```suggestion
find_duplicates <- function(x) {
```

File: R/assemble.R [hunk 4]
```diff
@@ -241,3 +252,55 @@ de_duplicate_names <- function(x) {
...
+  duplicated(dplyr::select_if(x, is.atomic))
+}
+
+# @param x a data.frame
+duplicate_drop = function(x) {
+  ix_dupes = find_duplicates(x)
```
Comment (mattroumaya) [inline R/assemble.R:272]:
```suggestion
duplicate_drop <- function(x) {
  ix_dupes <- find_duplicates(x)
```

File: R/assemble.R [hunk 5]
```diff
@@ -241,3 +252,55 @@ de_duplicate_names <- function(x) {
...
+    warning(
+      "There are duplicate responses, duplicates are dropped in
+      the results. Set fix_duplicates = 'keep' to retain them."
+    )
+  }
+  x[!ix_dupes, ]
+}
+
+# @param x a data.frame
+duplicate_keep = function(x) {
```
Comment (mattroumaya) [inline R/assemble.R:283]:
```suggestion
duplicate_keep <- function(x) {
```

File: R/assemble.R [hunk 6]
```diff
@@ -241,3 +252,55 @@ de_duplicate_names <- function(x) {
...
+    warning(
+      "There are duplicate responses, duplicates are retained in
+      the results. Set fix_duplicates = 'drop' to remove them."
+    )
+  }
+  x
+}
+
+# Deal with duplicate data by throwing an error
+duplicate_error = function(x) {
```
Comment (mattroumaya) [inline R/assemble.R:294]:
```suggestion
duplicate_error <- function(x) {
```

File: R/assemble.R [hunk 7]
```diff
@@ -241,3 +252,55 @@ de_duplicate_names <- function(x) {
...
+  # There should not be duplicate rows here, but putting this here in case of oddities like #27
+  assertthat::assert_that(!contains_duplicates(x),
+    msg = paste0(
+      "There are duplicated rows in the responses, maybe a situation like #27\n",
```
Comment (mattroumaya) [inline R/assemble.R:298]:
```suggestion
      "There are duplicated rows in the responses.\n",
```
Hoping this PR can close #27 and #74 for now. It seems that when thing start to leak through the API, there are new issues with duplicate rows, and likely aren't related to #27 anymore.

File: tests/testthat/test-assemble.R [hunk 8]
```diff
@@ -0,0 +1,26 @@
+no_dupes = cars[1, ]
+dupes = rbind(no_dupes, no_dupes)
```
Comment (mattroumaya) [inline tests/testthat/test-assemble.R:2]:
```suggestion
no_dupes <- cars[1, ]
dupes <- rbind(no_dupes, no_dupes)
```

File: tests/testthat/test-assemble.R [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+test_that("detect duplication", {
+  expect_true(contains_duplicates(dupes))
+  expect_false(contains_duplicates(no_dupes))
+})
+
+test_that("keep duplication", {
+  expect_warning(res <- duplicate_keep(dupes))
+  expect_identical(res, dupes)
+  expect_warning(res <- duplicate_keep(no_dupes), NA) # expect no warning
```
Comment (mattroumaya) [inline tests/testthat/test-assemble.R:12]: Could this be `expect_silent()` instead? No need to change but just a suggestion

File: R/assemble.R [hunk 10]
```diff
@@ -241,3 +252,55 @@ de_duplicate_names <- function(x) {
...
+  duplicated(dplyr::select_if(x, is.atomic))
+}
+
+# @param x a data.frame
+duplicate_drop = function(x) {
+  ix_dupes = find_duplicates(x)
+  if (sum(ix_dupes) > 0) {
+    warning(
+      "There are duplicate responses, duplicates are dropped in
+      the results. Set fix_duplicates = 'keep' to retain them."
```
Comment (mattroumaya) [inline R/assemble.R:276]:
```suggestion
      paste0("There are ", ix_dupes, " duplicate responses, duplicates are dropped in the results. Set fix_duplicates = 'keep' to retain them.")
```
Not necessary, but would be a good addition to alert the user about the # of duplicate responses that are returned.

**Review by mattroumaya — CHANGES_REQUESTED** [review]
Thanks so much for this PR! Just a few notes below: 
- To stay consistent, I think we should use `<-` as the assignment operator
- Probably makes sense to stop referring to #27 in warning messages at this point
- A more robust warning message about number of duplicates dropped might be useful

In the future, maybe an option to keep the newest or oldest record? I'm not able to test this with any survey I have access to, but it looks like the only option is to keep all or drop all duplicates. Definitely easy enough for someone to do that after pulling their survey data, but just throwing it out there. 

To merge this, please just fix the assignment operators as that's the only thing that I feel strongly about, otherwise it looks great.

File: tests/testthat/test-assemble.R (hunk 9 above)
Comment (jamieRowen) [reply tests/testthat/test-assemble.R:12]: As far as i understand `expect_silent()` expects no errors, warnings or messages so I deliberately opted for the expect no warning as that is really what I am expecting

File: R/assemble.R (hunk 10 above)
Comment (jamieRowen) [reply R/assemble.R]: Your suggested change didn't quite give the message you intended, have pushed a change that does

**Comment by jamieRowen** [comment]
@mattroumaya Changes made as requested

**Review by mattroumaya — APPROVED** [review]
